staging/src/k8s.io/dynamic-resource-allocation/controller/controller.go

```diff
@@ -517,50 +525,87 @@ func (ctrl *controller) getParameters(ctx context.Context, claim *resourcev1alph
 	return
 }
 
-func (ctrl *controller) allocateClaim(ctx context.Context,
-	claim *resourcev1alpha2.ResourceClaim, claimParameters interface{},
-	class *resourcev1alpha2.ResourceClass, classParameters interface{},
-	selectedNode string,
-	selectedUser *resourcev1alpha2.ResourceClaimConsumerReference) error {
+func (ctrl *controller) allocateClaims(ctx context.Context, claims claimAllocations, selectedNode string, selectedUser *resourcev1alpha2.ResourceClaimConsumerReference) error {
 	logger := klog.FromContext(ctx)
 
-	if claim.Status.Allocation != nil {
-		// This can happen when two PodSchedulingContext objects trigger
-		// allocation attempts (first one wins) or when we see the
-		// update of the PodSchedulingContext object.
-		logger.V(5).Info("Claim already allocated, nothing to do")
+	var needAllocation claimAllocations
+	for _, claim := range claims {
+		if claim.Claim.Status.Allocation != nil {
+			// This can happen when two PodSchedulingContext objects trigger
+			// allocation attempts (first one wins) or when we see the
+			// update of the PodSchedulingContext object.
+			logger.V(5).Info("Claim is already allocated, skipping allocation", "claim", claim.PodClaimName)
+			continue
+		}
+		needAllocation = append(needAllocation, claim)
+	}
+
+	if len(needAllocation) == 0 {
+		logger.V(5).Info("No claims need allocation, nothing to do")
 		return nil
 	}
 
-	claim = claim.DeepCopy()
-	if !ctrl.hasFinalizer(claim) {
-		// Set finalizer before doing anything. We continue with the updated claim.
-		logger.V(5).Info("Adding finalizer")
-		claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
-		var err error
-		claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
-		if err != nil {
-			return fmt.Errorf("add finalizer: %v", err)
+	for _, claimAllocation := range needAllocation {
+		if !ctrl.hasFinalizer(claimAllocation.Claim) {
+			patchFn := func() error {
+				claim := claimAllocation.Claim.DeepCopy()
+				// Set finalizer before doing anything. We continue with the updated claim.
+				logger.V(5).Info("Adding finalizer", "claim", claim.Name)
+				claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
+				var err error
+				claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
+				if err != nil {
+					return fmt.Errorf("add finalizer: %v", err)
+				}
+				ctrl.claimCache.Mutation(claim)
+				claimAllocation.Claim = claim
+				return nil
+			}
+			err := retry.RetryOnConflict(retry.DefaultRetry, patchFn)
+			if err != nil {
```

Comment:
This is not Kevin's code :grin: 

I, like the majority of the Go developers (?), prefer `if err := ...; err != nil`.

Reply:
I don't mind, I'll make it single line, but here's some stats:
~/git/kubernetes$ git grep "=.*RetryOnConflict(" | wc -l
69
~/git/kubernetes$ git grep "if.*=.*RetryOnConflict(" | wc -l
17

Which says that toal of 69 entried of "err = RetryOnConflict" or similar are there in k/k, and only 17 of them have if on the same line :)

Reply:
This check is flawed. Several of the lines which call RetryOnConflict outside of an `if` check are not followed by an if check and thus have to be excluded from the comparison.

The comparison has to be between `if err := ...; err != nil {...}` vs. `err := ...; if err != nil {...}`.